### PR TITLE
creates new tool, projectLink, and project entities

### PR DESCRIPTION
**Pull request by tdep**
In order to serve the front-end "**Projects**" section, we need to set up a new table to store the project information.
This should include the following:

**projects**
|column|type|attribute|
|---|---|---|
|id|int|primary key|
|title|String||
|description|String||
|projectLinks|List<ProjectLink>|One-To-Many|
|tools|Set<Tool>|Many-To-Many; Join Table|

There are also three other tables needed to fill in the data for project links and tools:

**project_links**
|column|type|attribute|
|---|---|---|
|id|int|primary key|
|project|Project|Many-To-One|
|name|String||
|url|String||

**tools**
|column|type|attribute|
|---|---|---|
|id|int|primary key|
|projects|Set<Project>|Many-To-Many; Join Table|
|name|String||

**project_tools (Join Table)**
|column|type|attribute|
|---|---|---|
|project_id|int|PK, FK1|
|tool_id|int|PK, FK2|

Each entity also needs associated `DAO`, `model`, and `repository` files.

### CODE REVIEW

File: src/main/java/com/tdep/tadlab/dao/ProjectDao.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.tdep.tadlab.dao;
+
+import com.tdep.tadlab.model.Project;
+import com.tdep.tadlab.model.Tool;
+
+import java.util.List;
+import java.util.Optional;
+
+public interface ProjectDao {
```
Comment (chaoms) [inline src/main/java/com/tdep/tadlab/dao/ProjectDao.java]: nit: based on vague memories from many years ago + brief skimming on google, if you're planning on leveraging Spring Data JPA, this should be something more like `interface ProjectRepository extends JpaRepository` 

Basically, if this class is intended to manage the actual DB queries, seems like the naming pattern is `repository` -- there's different solutions out of the box depending on how granular or basic your queries need to be, so you might want to look into that a bit.

File: src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package com.tdep.tadlab.repository;
+
+public class ProjectDaoImpl {
```
Comment (chaoms) [inline src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java]: Some suggestions for moving files around:
- Have the `ProjectRepository`, `*Repository`, etc files go into this `/repository` directory. 
  - I don't think there's a need to have a `DAO` and a `DAOImpl` (?) Unless you're following some other template pattern -- but I think generally, you can just declare `interface ProjectRepository extends JpaRepository` (or extending whichever Spring Data JPA repository makes the most sense for you), and the queries can live directly on the interface
- I don't think you need the `/dao` directory if you end up following the above ^ 
- May want to consider making a `/dto` directory for storing your DTO classes -- these generally represent the request structure coming from the client, AKA what you're receiving from the UI. Most of the times, in the `service` layer, there's some kind of conversion from DTO to the DB model

File: src/main/java/com/tdep/tadlab/dao/ProjectDao.java (hunk 1 above)
Comment (tdep) [reply src/main/java/com/tdep/tadlab/dao/ProjectDao.java]: I think you're right - that is a pattern I've been coming across as I look things up. I think I was trying to originally use a DAO structure but sort of stumbled into Spring Boot and found a way to sort of use both? I'll try this with the Project service and see if I'll extend it to the existing Job service.

File: src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java (hunk 2 above)
Comment (tdep) [reply src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java]: This also makes sense to me. Though I do wonder, is DTO basically synonymous with 'Controller' ?

Comment (chaoms) [reply src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java]: Ya, I think you could make the argument that DTO objects basically only life at the controller level (incoming requests are DTO, controller responses are also DTO)

File: src/main/java/com/tdep/tadlab/dao/ProjectDao.java (hunk 1 above)
Comment (chaoms) [reply src/main/java/com/tdep/tadlab/dao/ProjectDao.java]: I think the DAO pattern is also legit, but maybe less favored than the repository approach (me spitballing based off the fact that the official Spring tutorials seem to all reference the repository pattern). I think Spring tried to make the basic CRUD operations as out-of-the-box as possible so the repository approach gives you a lot of free queries right off the bat

Comment (tdep) [reply src/main/java/com/tdep/tadlab/dao/ProjectDao.java]: Agreed, I've set up the new service to follow the repository pattern and will create the new tables to match when I merge this pr. Fingers crossed!

File: src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java (hunk 2 above)
Comment (tdep) [reply src/main/java/com/tdep/tadlab/repository/ProjectDaoImpl.java]: Following the guides, it seems they like to have the following:

- `/entity`
- `/repository`
- `/service`
- `/controller`

Where the service package has both `*Service` and `*ServiceImpl` which is then used by the controllers.

**Pull request merged**